ykbh/src/lib.rs

```diff
@@ -0,0 +1,267 @@
+use std::alloc::{alloc, dealloc, Layout};
+use std::convert::{TryFrom, TryInto};
+use ykpack::{
+    self, Constant, ConstantInt, IPlace, Local, LocalDecl, Statement, Terminator, TypeId,
+    UnsignedInt,
+};
+use yktrace::sir::SIR;
+
+const SUSIZE: usize = std::mem::size_of::<usize>();
+
+pub struct SIRInterpreter {
+    map: *mut u8,
+    layout: Layout,
+    locals: Vec<usize>,
+}
+
+impl SIRInterpreter {
+    pub fn new(local_decls: &Vec<LocalDecl>) -> Self {
+        // FIXME Soon this will be pre-computed and handed to us by SIR.
+        let mut size: usize = 0;
+        let mut locals = Vec::new();
+        for d in local_decls {
+            locals.push(size);
+            size += usize::try_from(SIR.ty(&d.ty).size()).unwrap();
+        }
+
+        // Allocate memory for the locals
+        let layout = Layout::from_size_align(usize::try_from(size).unwrap(), 8).unwrap();
+        let map = unsafe { alloc(layout) };
+
+        SIRInterpreter {
+            map,
+            layout,
+            locals,
+        }
+    }
+
+    /// Sets the current frame of the interpreter.
+    pub fn set_frame(&mut self, tio: usize) {
```

Comment:
I'm not sure what `tio` does here, in particular why we we might be calling `to_ne_bytes` on it. Any clues?

Reply:
`tio` is a pointer to the trace inputs which I need to store inside the locals map. This will go away (hopefully) once I get rid of the pointer as usize stuff.

Reply:
Okay, this is a bit more complicated. We need `to_ne_bytes` as we are interested in the value of pointer itself, not what it points to. So we need to convert the `&IO` into a slice of bytes or a `*const u8`. If there's a better way than going through a `usize` I'm all ears.

Reply:
I don't know what this function is really intending to do, so it's very hard for me to say anything sensible I'm afraid.

Reply:
At the moment, all it does is pass the pointer to the `trace inputs` to the interpreter. `Trace inputs` is what we also hand to the compiler, which contains stuff like the program counter, etc. If you look at the tests below you can see how `tio` is used.

Reply:
I guess what I'm saying is: I don't understand the high-level structure that leads to this decision. My default reaction is that this code is unlikely to be correct under stack borrows, but perhaps it is, and perhaps this is the best way of doing things. I don't know.

Reply:
Since we are not writing to `tio` until `interpret` returns, this should be fine under stacked borrows. To be sure I've tested this in MIRI which seems happy: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=8a3fd007a16b3db15720382f9b486611

I've added a small comment in 4f9d18c and changed the argument type from `usize` to `*mut usize` to avoid confusion in 1a5048d.